### DISCUSSION TITLE
How to validate database inserts/updates? & Is it possible to use Supabase without login?

### DISCUSSION THREAD

**Discussion by panmona** [supabase/supabase#1980] [Questions] [answered]
Some question came  up while I was evaluating whether Supabase would be a good fit for adding an online mode to a [turn based game](https://playtictactoe.xyz) I made.
To better understand what I want to build, here's a short description on how the online mode would work:
1. User A creates game, receives Id
2. User A shares Id link with User B
3. User B uses the link with Id and joins the Game
4. User A makes one move
5. User B receives the move by the server and is now able to make a move himself

Here are my questions ordered by importance:
1. I think the general workflow would be quite fitting to use Supabase for or do you see something that will be difficult/not possible?
2. How would I go about validating that the moves user A & B make are valid according to the game rules on the server before they are inserted? Do I need to wait for Function functionality for this? What is the timeline on that / where can I find more info?
3. Is it possible to use Supabase without requiring users to login but still only allow them to access the records of the game they're currently playing? (this means the game with that specific Id) 
Or maybe at least only allow them to query by a specific Id, then it's similar to validating the user id.
I only found examples that use auth. This is not a must have but it would be nice to not require a login.
4. If it is not possible to use it without Auth according to my requirements: How would I go about only allowing User A and B to access the data related to a specific game?

I hope someone will be able to answer these questions, thanks a lot in advance :bowing_man:
Looking forward to hopefully be able to use Supabase for this project :100: :rocket:

**Comment by silentworks** [comment]
1. This shouldn't be a problem at all for you to use Supabase, I'm working on something similar but I decided to go with user login for mine.
2. You can use functions in the database to validate something before it inserting it into the database, however, I would like to think the game rules would sit client-side rather than in a database function.
3. Yes this is possible, it would require you to create unique links for each game that someone can't easily guess, but if someone was to figure this out there wouldn't be a way to stop them from viewing that game or joining it.
4. You could create a pivot table that has the association between User A and B to the game or you can create a participants table that keeps a reference to the user_id and game_id.

I'm planning on making my game open source and will share the code on here once I have completed it. But I'm still far off from completion.

**Reply by panmona** [reply]
1. Good :+1: 
2. I don't think this will work well in a database function but I have never used Postgres for this before. 
Yes they are also 'enforced' client-side but the client can be changed so this isn't secure. So this would just be used to prevent cheating.
I still feel like this would require functions, maybe someone else can chime in here?
3. Sounds reasonable. Not sure though how I can validate that User A doesn't move before User B as I don't have any User Ids.
4. Yes something like that is a possibility but that won't keep them from accessing different data. If I go with auth, I want to at least have short game ids that could also be guessed.

**Reply by silentworks** [reply]
2. Can you give an example of what these rules would look like?
3. You would definitely have to create a user in your own table, doesn't have to use the Supabase auth stuff. This would mean you have a unique identifier for each user, otherwise, there is no way to have a unique id.
4. Yeah I'm using short game ids with auth in my case. I just set a constraint in the database to make sure that the game code is unique. Without the auth system, it will be hard to block others from accessing a game, because the tables would need row level security in order to keep them secure.

**Reply by panmona** [reply]
2. A user is only allowed to make a move in only a few specific fields on a 9x9 board. To validate this the current move needs to be checked against the current state of the board which can be built from all previous moves.
3. My question here probably boils down to where I would best generate these ids. (another case where I would reach for a function) I don't think that should be done from the client.
4. I still don't quite get how this would be done. I probably need to read much more about how RLS can be implemented.

**Reply by panmona** [reply]
About 3:
Isn't it always possible to read the whole table when RTS is not enabled? So how can this be secured?

**Comment by panmona** [comment, marked as answer]
I'm still not sure what the best way would be to verify the database inserts and whether that will be possible with the upcoming functions.
I don't think implementing the logic in a stored procedure is feasible. Should I maybe make the inserts from an external function and then migrate to Supabase functions later?

**Reply by silentworks** [reply]
Yes you can do this for now and migrate later on.
